tikv/region.go

```diff
@@ -141,46 +142,60 @@ func (ri *regionCtx) marshal() []byte {
 	return data
 }
 
-func (ri *regionCtx) tryAcquireLatches(hashVals []uint64) (bool, *sync.WaitGroup) {
+func (ri *regionCtx) tryAcquireLatches(hashVals []uint64) {
 	wg := new(sync.WaitGroup)
 	wg.Add(1)
-	ri.latchesMu.Lock()
-	defer ri.latchesMu.Unlock()
 	for _, hashVal := range hashVals {
-		if wg, ok := ri.latches[hashVal]; ok {
-			return false, wg
+		for {
+			res, loaded := ri.latches.LoadOrStore(hashVal, wg)
+			if loaded {
+				resWg := res.(*sync.WaitGroup)
+				resWg.Wait()
+				continue
+			}
+			break
 		}
 	}
-	for _, hashVal := range hashVals {
-		ri.latches[hashVal] = wg
-	}
-	return true, nil
 }
 
+// AcquireLatches add latches for all input hashVals, the input hashVals should be
+// sorted and have no duplicates
 func (ri *regionCtx) AcquireLatches(hashVals []uint64) {
-	start := time.Now()
-	for {
-		ok, wg := ri.tryAcquireLatches(hashVals)
-		if ok {
-			dur := time.Since(start)
-			metrics.LatchWait.Observe(dur.Seconds())
-			if dur > time.Millisecond*50 {
-				log.Warnf("region %d acquire %d locks takes %v", ri.meta.Id, len(hashVals), dur)
-			}
-			return
+	if len(hashVals) == 0 {
+		log.Errorf("hashVals array length equals 0")
+		return
+	}
+	// Make sure hashVals in sort ascending order and no duplicates
```

Comment:
We better do this in `keysToHashVals`.
The `ReleaseLatches` use the original hashes that may contain duplicated hashVal.